### ISSUE TITLE
Problem with startLife > 1 in qsdConverge

### ISSUE THREAD

**Issue by patrickbarks** [jonesor/Rage#52]
I think there's an issue with using `startLife > 1` in `qsdConverge`. Because the vast majority of matU are reducible (i.e. b/c there are no repro transitions to 'complete the cycle'), convergence to the quasi-stable distribution may depend on the initial population vector, which is a function of `startLife`. E.g. in the matrix below, because the stable distribution of the 1st stage is >> 0, and there is no retrogression to stage 1, we would never get convergence with any `startLife > 1`.

```
matU <- rbind(c(0.5, 0.0, 0.0, 0.0),
              c(0.1, 0.3, 0.1, 0.0),
              c(0.0, 0.1, 0.2, 0.1),
              c(0.0, 0.0, 0.1, 0.3))

popbio::stable.stage(matU)
## 0.5000 0.3125 0.1250 0.0625

popdemo::isErgodic(matU)
## FALSE
```
I'm not sure how to deal with this. Checking `isErgodic(matU)` could help, but there are non-ergodic matU that still do converge to the QSD (for some startLife), and a few ergodic but imprimitive matU that don't converge.

**Comment by patrickbarks** [comment]
Here's some explanation for my proposed changes to `qsdConverge` in #51:

The purpose of removing stages that can never be reached _from_ stage `startLife` is to make sure that the 'target stage distribution' that we use to assess convergence is in fact the stage distribution that our population vector actually converges to (for a given value of `startLife`).

In `qsdConverge`, we populate the stage vector with a single individual at stage `startLife`, and then repeatedly project through `matU` (equivalent to the procedure in `mpm_to_lx`, `makeLifeTable`, and `longevity`). Any stage that has no connection _from_ stage `startLife` has no impact on  the actual distribution that our starting population vector will converge to (nor do such stages have any impact on our measures of `lx`, `longevity`, or `lifeExpectancy`, from stage `startLife`). So I don't think there's any problem with removing these stages.

**Example**

Here's an example non-ergodic `matU` where, for `startLife == 1` the observed final stage distribution matches `stable.stage(matU)`, but for `startLife == 2` it does not (though the population vector still does converge to a final distribution).

```
library(popbio); library(popdemo)

matU1 <- rbind(c(0.5, 0.0, 0.0, 0.0),
               c(0.1, 0.3, 0.1, 0.0),
               c(0.0, 0.1, 0.2, 0.1),
               c(0.0, 0.0, 0.1, 0.2))

popdemo::isErgodic(matU1)
>> [1] FALSE

popbio::stable.stage(matU1)
>> [1] 0.52 0.32 0.12 0.04
```
Observed final stage distribution for `startLife == 1`...
```
# stage vec for startLife == 1
n <- c(1, 0, 0, 0)

# project n through matU1 for 500 time steps
proj <- project(matU1, vector = n, time = 500)@vec

# observed stage dist at times 499:500 (matches stable.stage)
apply(proj, 1, function(x) x / sum(x))[,499:500]
>>    [,1] [,2]
>> S1 0.52 0.52
>> S2 0.32 0.32
>> S3 0.12 0.12
>> S4 0.04 0.04
```
Observed final stage distribution for `startLife == 2`...
```
# stage vec for startLife == 2
n <- c(0, 1, 0, 0)

# project n through matU1 for 500 time steps
proj <- project(matU1, vector = n, time = 500)@vec

# observed stage dist at times 499:500 (doesn't match stable.stage)
apply(proj, 1, function(x) x / sum(x))[,499:500]
>>         [,1]      [,2]
>> S1 0.0000000 0.0000000
>> S2 0.4450419 0.4450419
>> S3 0.3568959 0.3568959
>> S4 0.1980623 0.1980623
```
The problem for `startLife == 2` is that the observed final distribution doesn't match the stable distribution (and so will appear to never converge based on the previous `qsdConverge`), which is possible because `matU1` isn't ergodic. If we remove the stages with no connection _from_ `startLife` (and update `startLife` accordingly), none of our survival-related metrics will be affected.

```
# remove stage with no connections from stage 2
matU2 <- matU1[-1,-1]

mpm_to_lx(matU1, startLife = 2, N = 5) == 
 mpm_to_lx(matU2, startLife = 1, N = 5)
>> [1] TRUE TRUE TRUE TRUE TRUE TRUE

lifeExpectancy(matU1, startLife = 2) == 
 lifeExpectancy(matU2, startLife = 1)
>> [1] TRUE

longevity(matU1, startLife = 2) == 
 longevity(matU2, startLife = 1)
>> [1] TRUE
```

But we'll now get a sensible value from `qsdConverge` because `matU2` is ergodic, and so the target distribution (`stable.stage(matU2)`) matches the distribution that our starting population vector will actually converge to.
```
# (these examples use the old qsdConverge function)
qsdConverge(matU1, startLife = 2)
>> [1] NA
>> Warning: Convergence not reached within nSteps

qsdConverge(matU2, startLife = 1)
>> [1] 6
```

**Comment by patrickbarks** [comment]
I think we should remove the `ergodicFix` argument introduced to `qsd_converge` in #51, and simply add documentation to clarify how we calculate the observed stable distribution given `matU` and `startLife`.

Because QSD is dependent on a given value of `startLife`, _it is not fundamentally dependent on ergodicity_. Ergodicity is just a convenience that allows us to calculate the observed stable distribution as the dominant right eigenvector.

In short, I see no reason to ever set `ergodicFix = FALSE` and thereby exclude non-ergodic `matU` from the calculation of QSD. If a user for some reason wants to exclude non-ergodic `matU` from their analysis (recall about 1/3 of mean matU in COMPADRE are non-ergodic), they should use `popdemo::isErgodic` rather than relying on an error message from `qsd_converge`.

**Comment by patrickbarks** [comment]
One remaining problem for `qsd_converge` is any age-based matU _without_ a stasis loop at the maximum age class.

These will always be non-ergodic, and the eigenvector-based stable distribution (e.g. from `popbio::stable.stage`) will generally be a vector of 0s with a single 1 at the maximum age. However, the empirical stable distribution obtained via projection analysis will generally just be a vector of 0s (because without a stasis loop, no individuals remain after the maxmimum age class). E.g.
```
# age-based matU
matU <- rbind(c(0.0, 0.0, 0.0, 0.0),
              c(0.5, 0.0, 0.0, 0.0),
              c(0.0, 0.4, 0.0, 0.0),
              c(0.0, 0.0, 0.3, 0.0))

# throws error
qsd_converge(matU)
```
Not sure exactly how we should deal with these.

**Comment by jonesor** [comment]
Closing this now.
